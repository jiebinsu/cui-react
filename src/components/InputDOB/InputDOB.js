import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";

const InputDOB = ({ id, name, label, hint, hasError, errors, refs }) => {
  const dayClasses = classNames(
    "coop-form__field coop-form__input coop-form__input--width-2",
    [hasError && "coop-form__invalid"]
  );
  const monthClasses = classNames(
    "coop-form__field coop-form__input coop-form__input--width-2",
    [hasError && "coop-form__invalid"]
  );
  const yearClasses = classNames(
    "coop-form__field coop-form__input coop-form__input--width-4",
    [hasError && "coop-form__invalid"]
  );

  return (
    <div className="coop-form__row">
      <fieldset aria-describedby={`${id}-hint ${id}-error`}>
        {label && <legend className="coop-form__label">{label}</legend>}
        {hint && (
          <p id={`${id}-hint`} className="coop-form__hint">
            {hint}
          </p>
        )}
        {hasError &&
          Object.entries(errors).map(([key, value]) => (
            <p id={`${id}-${key}-error`} key={key} className="coop-form__error">
              {value}
            </p>
          ))}

        <div className="coop-form__columns">
          <div className="coop-form__column">
            <label className="coop-form__label" htmlFor={`${id}-day`}>
              Day
            </label>
            <input
              id={`${id}-day`}
              name={`${name}-day`}
              type="text"
              pattern="[0-9]*"
              inputMode="numeric"
              maxLength="2"
              minLength="2"
              autoComplete="bday-day"
              className={dayClasses}
              ref={refs.day}
            />
          </div>
          <div className="coop-form__column">
            <label className="coop-form__label" htmlFor={`${id}-month`}>
              Month
            </label>
            <input
              id={`${id}-month`}
              name={`${name}-month`}
              type="text"
              pattern="[0-9]*"
              inputMode="numeric"
              maxLength="2"
              minLength="2"
              autoComplete="bday-month"
              className={monthClasses}
              ref={refs.month}
            />
          </div>
          <div className="coop-form__column">
            <label className="coop-form__label" htmlFor={`${id}-year`}>
              Year
            </label>
            <input
              id={`${id}-year`}
              name={`${name}-year`}
              type="text"
              pattern="[0-9]*"
              inputMode="numeric"
              maxLength="4"
              minLength="4"
              autoComplete="bday-year"
              className={yearClasses}
              ref={refs.year}
            />
          </div>
        </div>
      </fieldset>
    </div>
  );
};

InputDOB.defaultProps = {
  label: "What is your date of birth?",
  hint: "For example 21 12 2020",
  hasError: false,
  errors: {
    default: "The date of birth you entered is invalid",
  },
  refs: {},
};

InputDOB.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  hint: PropTypes.string,
  hasError: PropTypes.bool,
  errors: PropTypes.shape({
    day: PropTypes.string,
    month: PropTypes.string,
    year: PropTypes.string,
  }),
  refs: PropTypes.shape({
    day: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.shape({ current: PropTypes.elementType }),
    ]),
    month: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.shape({ current: PropTypes.elementType }),
    ]),
    year: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.shape({ current: PropTypes.elementType }),
    ]),
  }),
};

export default InputDOB;
