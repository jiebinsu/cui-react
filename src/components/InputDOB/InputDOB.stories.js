import React from "react";
import InputDOB from "./InputDOB";

export default {
  title: "Base/InputDOB",
  component: InputDOB,
};

export const Basic = (args) => <InputDOB {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  label: "Date of Birth",
};

const errors = {
  day: "The day you entered is invalid",
  month: "The month you entered is invalid",
  year: "The year you entered is invalid",
};
export const WithError = () => (
  <InputDOB
    id="withError"
    name="withError"
    label="With Error"
    errors={errors}
    hasError
  />
);
