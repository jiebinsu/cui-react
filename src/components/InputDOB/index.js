import InputDOB from "./InputDOB";

export default InputDOB;
