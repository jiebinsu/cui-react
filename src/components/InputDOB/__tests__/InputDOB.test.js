import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputDOB from "../index";

describe("InputDOB", () => {
  it("should render correctly", () => {
    const wrapper = mount(<InputDOB id="test" name="test" />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer
      .create(
        <InputDOB
          id="test"
          name="test"
          label="Date of Birth"
          hint="01 01 1990"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render with errors correctly", () => {
    const errors = {
      day: "The day you entered is invalid",
      month: "The month you entered is invalid",
      year: "The year you entered is invalid",
    };
    const wrapper = mount(
      <InputDOB id="test" name="test" hasError errors={errors} />
    );
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer.create(<InputDOB id="test" name="test" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
