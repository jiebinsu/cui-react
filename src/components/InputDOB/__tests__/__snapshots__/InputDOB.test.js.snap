// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDOB should render correctly 1`] = `
<div
  className="coop-form__row"
>
  <fieldset
    aria-describedby="test-hint test-error"
  >
    <legend
      className="coop-form__label"
    >
      Date of Birth
    </legend>
    <p
      className="coop-form__hint"
      id="test-hint"
    >
      01 01 1990
    </p>
    <div
      className="coop-form__columns"
    >
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-day"
        >
          Day
        </label>
        <input
          autoComplete="bday-day"
          className="coop-form__field coop-form__input coop-form__input--width-2"
          id="test-day"
          inputMode="numeric"
          maxLength="2"
          minLength="2"
          name="test-day"
          pattern="[0-9]*"
          type="text"
        />
      </div>
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-month"
        >
          Month
        </label>
        <input
          autoComplete="bday-month"
          className="coop-form__field coop-form__input coop-form__input--width-2"
          id="test-month"
          inputMode="numeric"
          maxLength="2"
          minLength="2"
          name="test-month"
          pattern="[0-9]*"
          type="text"
        />
      </div>
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-year"
        >
          Year
        </label>
        <input
          autoComplete="bday-year"
          className="coop-form__field coop-form__input coop-form__input--width-4"
          id="test-year"
          inputMode="numeric"
          maxLength="4"
          minLength="4"
          name="test-year"
          pattern="[0-9]*"
          type="text"
        />
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`InputDOB should render with errors correctly 1`] = `
<div
  className="coop-form__row"
>
  <fieldset
    aria-describedby="test-hint test-error"
  >
    <legend
      className="coop-form__label"
    >
      What is your date of birth?
    </legend>
    <p
      className="coop-form__hint"
      id="test-hint"
    >
      For example 21 12 2020
    </p>
    <div
      className="coop-form__columns"
    >
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-day"
        >
          Day
        </label>
        <input
          autoComplete="bday-day"
          className="coop-form__field coop-form__input coop-form__input--width-2"
          id="test-day"
          inputMode="numeric"
          maxLength="2"
          minLength="2"
          name="test-day"
          pattern="[0-9]*"
          type="text"
        />
      </div>
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-month"
        >
          Month
        </label>
        <input
          autoComplete="bday-month"
          className="coop-form__field coop-form__input coop-form__input--width-2"
          id="test-month"
          inputMode="numeric"
          maxLength="2"
          minLength="2"
          name="test-month"
          pattern="[0-9]*"
          type="text"
        />
      </div>
      <div
        className="coop-form__column"
      >
        <label
          className="coop-form__label"
          htmlFor="test-year"
        >
          Year
        </label>
        <input
          autoComplete="bday-year"
          className="coop-form__field coop-form__input coop-form__input--width-4"
          id="test-year"
          inputMode="numeric"
          maxLength="4"
          minLength="4"
          name="test-year"
          pattern="[0-9]*"
          type="text"
        />
      </div>
    </div>
  </fieldset>
</div>
`;
